[MSVC] When building Google_Protobuf emits error C2338 on MSVC

Hi, all
Google_Protobuf fails to build on MSVC build main_20220223.01 due to error C2338. Could you please help look at this issue? Thank you very much.

**What version of protobuf and what language are you using?**
Version: 29b3d01

**What operating system (Linux, Windows, ...) and version?**
Window 10

**What runtime / compiler are you using (e.g., python version or gcc version)**
git 

**What did you do?**
Steps to reproduce the behavior:
1. open VS2019 x64 tools command
2. git clone https://github.com/google/protobuf F:\gitP\google\protobuf\build_x86
3. set VSCMD_SKIP_SENDTELEMETRY=1 & "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat" -host_arch=x86 -arch=x86
4. set originalLIB=%_LIB_%
5. set _LIB_=/MACHINE:X86 %orignalLIB%
6. cd F:\gitP\google\protobuf\build_x86
7. msbuild /p:Platform=Win32 /p:Configuration=Release protobuf.sln /t:Rebuild 2>&1

**What did you expect to see**
Compile fine.
**What did you see instead?**
F:\gitP\google\protobuf\src\google/protobuf/repeated_field.h(81,27): error C2338: static_assert failed: '' (compiling source file F:\gitP\google\protobuf\src\google\protobuf\extension_set_unittest.cc) [F:\gitP\google\protobuf\build_x86\tests.vcxproj]
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Build log**

[Google build.txt](https://github.com/protocolbuffers/protobuf/files/8175871/Google.build.txt)

Hi @deannagarcia 
I met the same issue. The commit of protobuf we use now is c38281dd20e562bac239bc77ab2fa10f71661708
The Visual Studio version is vs2019(VS16.11.11), OS is Windows Server 2019, it reports error C2338 on x86 mode. 
You can check the details in the attached log file. Could you please help look this issue? Thanks.
[build.log](https://github.com/protocolbuffers/protobuf/files/9117331/build.log)

Yes, only `repeated_field.h` file emits an error when compiling `extension_set_unittest.cc` file.

`F:\gitP\google\protobuf\src\google/protobuf/repeated_field.h(81,27): error C2338: static_assert failed: '' (compiling source file F:\gitP\google\protobuf\src\google\protobuf\extension_set_unittest.cc)`

--- a/kokoro/linux/32-bit/test_php.sh
+++ b/kokoro/linux/32-bit/test_php.sh
@@ -36,10 +36,16 @@ build_php_c() {
   test_php_c
 }
 
-cmake .
+mkdir build
+pushd build
+cmake ..
 cmake --build . -- -j20
 ctest --verbose --parallel 20
 export PROTOC=$(pwd)/protoc
+popd
+
+# Cleanup after Bazel build
+rm -rf build
 
 build_php 7.0
 build_php 7.1

--- a/kokoro/linux/32-bit/test_php.sh
+++ b/kokoro/linux/32-bit/test_php.sh
@@ -38,6 +38,7 @@ build_php_c() {
 
 cmake .
 cmake --build . -- -j20
+ctest --verbose --parallel 20
 export PROTOC=$(pwd)/protoc
 
 build_php 7.0

--- a/kokoro/linux/32-bit/test_php.sh
+++ b/kokoro/linux/32-bit/test_php.sh
@@ -37,7 +37,7 @@ build_php_c() {
 }
 
 cmake .
-cmake --build . --target protoc -- -j20
+cmake --build . -- -j20
 export PROTOC=$(pwd)/protoc
 
 build_php 7.0

--- a/kokoro/linux/32-bit/test_php.sh
+++ b/kokoro/linux/32-bit/test_php.sh
@@ -41,10 +41,11 @@ pushd build
 cmake ..
 cmake --build . -- -j20
 ctest --verbose --parallel 20
-export PROTOC=$(pwd)/protoc
 popd
 
 # Cleanup after Bazel build
+mv build/protoc .
+export PROTOC=$(pwd)/protoc
 rm -rf build
 
 build_php 7.0

--- a/kokoro/linux/32-bit/test_php.sh
+++ b/kokoro/linux/32-bit/test_php.sh
@@ -41,12 +41,8 @@ pushd build
 cmake ..
 cmake --build . -- -j20
 ctest --verbose --parallel 20
-popd
-
-# Cleanup after Bazel build
-mv build/protoc .
 export PROTOC=$(pwd)/protoc
-rm -rf build
+popd
 
 build_php 7.0
 build_php 7.1
@@ -57,3 +53,6 @@ build_php_c 7.4
 build_php_c 7.1-zts
 build_php_c 7.2-zts
 build_php_c 7.5-zts
+
+# Cleanup after CMake build
+rm -rf build

--- a/php/generate_descriptor_protos.sh
+++ b/php/generate_descriptor_protos.sh
@@ -5,7 +5,9 @@
 
 set -e
 
-PROTOC=$(realpath protoc)
+if [[ -z "${PROTOC}" ]]; then
+  PROTOC=$(realpath protoc)
+fi
 if [ ! -f $PROTOC ]; then
   bazel build -c opt //:protoc
   PROTOC=$(realpath bazel-bin/protoc)

--- a/php/generate_test_protos.sh
+++ b/php/generate_test_protos.sh
@@ -4,7 +4,9 @@ set -ex
 
 cd `dirname $0`/..
 
-PROTOC=$(pwd)/protoc
+if [[ -z "${PROTOC}" ]]; then
+  PROTOC=$(pwd)/protoc
+fi
 if [ ! -f $PROTOC ]; then
   bazel build -c opt //:protoc
   PROTOC=$(pwd)/bazel-bin/protoc

--- a/src/google/protobuf/compiler/cpp/message_size_unittest.cc
+++ b/src/google/protobuf/compiler/cpp/message_size_unittest.cc
@@ -139,9 +139,9 @@ TEST(GeneratedMessageTest, OneStringSize) {
 
 TEST(GeneratedMessageTest, MoreStringSize) {
   struct MockGenerated : public MockMessageBase {  // 16 bytes
-    int has_bits[1];                               // 4 bytes
     int cached_size;                               // 4 bytes
     MockRepeatedPtrField data;                     // 24 bytes
+    // + 4 bytes padding
   };
   GOOGLE_CHECK_MESSAGE_SIZE(MockGenerated, 48);
   EXPECT_EQ(sizeof(protobuf_unittest::MoreString), sizeof(MockGenerated));

--- a/src/google/protobuf/extension_set_unittest.cc
+++ b/src/google/protobuf/extension_set_unittest.cc
@@ -855,8 +855,10 @@ TEST(ExtensionSetTest, SpaceUsedExcludingSelf) {
     const size_t old_capacity =                                                \
         message->GetRepeatedExtension(unittest::repeated_##type##_extension)   \
             .Capacity();                                                       \
-    EXPECT_GE(old_capacity,                                                    \
-              (RepeatedFieldLowerClampLimit<cpptype, sizeof(void*)>()));       \
+    EXPECT_GE(                                                                 \
+        old_capacity,                                                          \
+        (RepeatedFieldLowerClampLimit<cpptype, std::max(sizeof(cpptype),       \
+                                                        sizeof(void*))>()));   \
     for (int i = 0; i < 16; ++i) {                                             \
       message->AddExtension(unittest::repeated_##type##_extension, value);     \
     }                                                                          \

--- a/src/google/protobuf/io/zero_copy_stream_unittest.cc
+++ b/src/google/protobuf/io/zero_copy_stream_unittest.cc
@@ -720,6 +720,9 @@ TEST_F(IoTest, StringIo) {
 
 // Verifies that outputs up to kint32max can be created.
 TEST_F(IoTest, LargeOutput) {
+  // Filter out this test on 32-bit architectures.
+  if(sizeof(void*) < 8) return;
+
   std::string str;
   StringOutputStream output(&str);
   void* unused_data;

--- a/src/google/protobuf/repeated_field_unittest.cc
+++ b/src/google/protobuf/repeated_field_unittest.cc
@@ -429,14 +429,14 @@ TEST(RepeatedField, ReserveNothing) {
 
 TEST(RepeatedField, ReserveLowerClamp) {
   int clamped_value = internal::CalculateReserveSize<bool, sizeof(void*)>(0, 1);
-  EXPECT_GE(clamped_value, 8 / sizeof(bool));
+  EXPECT_GE(clamped_value, sizeof(void*) / sizeof(bool));
   EXPECT_EQ((internal::RepeatedFieldLowerClampLimit<bool, sizeof(void*)>()),
             clamped_value);
   // EXPECT_EQ(clamped_value, (internal::CalculateReserveSize<bool,
   // sizeof(void*)>( clamped_value, 2)));
 
   clamped_value = internal::CalculateReserveSize<int, sizeof(void*)>(0, 1);
-  EXPECT_GE(clamped_value, 8 / sizeof(int));
+  EXPECT_GE(clamped_value, sizeof(void*) / sizeof(int));
   EXPECT_EQ((internal::RepeatedFieldLowerClampLimit<int, sizeof(void*)>()),
             clamped_value);
   // EXPECT_EQ(clamped_value, (internal::CalculateReserveSize<int,

--- a/src/google/protobuf/util/time_util_test.cc
+++ b/src/google/protobuf/util/time_util_test.cc
@@ -48,15 +48,18 @@ using google::protobuf::Timestamp;
 namespace {
 
 TEST(TimeUtilTest, TimestampStringFormat) {
-  Timestamp begin, end;
-  EXPECT_TRUE(TimeUtil::FromString("0001-01-01T00:00:00Z", &begin));
-  EXPECT_EQ(TimeUtil::kTimestampMinSeconds, begin.seconds());
-  EXPECT_EQ(0, begin.nanos());
-  EXPECT_TRUE(TimeUtil::FromString("9999-12-31T23:59:59.999999999Z", &end));
-  EXPECT_EQ(TimeUtil::kTimestampMaxSeconds, end.seconds());
-  EXPECT_EQ(999999999, end.nanos());
-  EXPECT_EQ("0001-01-01T00:00:00Z", TimeUtil::ToString(begin));
-  EXPECT_EQ("9999-12-31T23:59:59.999999999Z", TimeUtil::ToString(end));
+  // These these are out of bounds for 32-bit architectures.
+  if(sizeof(time_t) >= sizeof(uint64_t)) {
+    Timestamp begin, end;
+    EXPECT_TRUE(TimeUtil::FromString("0001-01-01T00:00:00Z", &begin));
+    EXPECT_EQ(TimeUtil::kTimestampMinSeconds, begin.seconds());
+    EXPECT_EQ(0, begin.nanos());
+    EXPECT_TRUE(TimeUtil::FromString("9999-12-31T23:59:59.999999999Z", &end));
+    EXPECT_EQ(TimeUtil::kTimestampMaxSeconds, end.seconds());
+    EXPECT_EQ(999999999, end.nanos());
+    EXPECT_EQ("0001-01-01T00:00:00Z", TimeUtil::ToString(begin));
+    EXPECT_EQ("9999-12-31T23:59:59.999999999Z", TimeUtil::ToString(end));
+  }
 
   // Test negative timestamps.
   Timestamp time = TimeUtil::NanosecondsToTimestamp(-1);
@@ -94,9 +97,12 @@ TEST(TimeUtilTest, DurationStringFormat) {
   EXPECT_TRUE(TimeUtil::FromString("0001-01-01T00:00:00Z", &begin));
   EXPECT_TRUE(TimeUtil::FromString("9999-12-31T23:59:59.999999999Z", &end));
 
-  EXPECT_EQ("315537897599.999999999s", TimeUtil::ToString(end - begin));
+  // These these are out of bounds for 32-bit architectures.
+  if(sizeof(time_t) >= sizeof(uint64_t)) {
+    EXPECT_EQ("315537897599.999999999s", TimeUtil::ToString(end - begin));
+    EXPECT_EQ("-315537897599.999999999s", TimeUtil::ToString(begin - end));
+  }
   EXPECT_EQ(999999999, (end - begin).nanos());
-  EXPECT_EQ("-315537897599.999999999s", TimeUtil::ToString(begin - end));
   EXPECT_EQ(-999999999, (begin - end).nanos());
 
   // Generated output should contain 3, 6, or 9 fractional digits.

